File: ruby_and_python/13_search/binary_search.py
# Write a Python program for binary search.
# Binary Search: I computer science, a biary search or half-interval search algorithm fins the position of a target value
# within a sorted array. The binary search algorithm can be clasifiend as a dichotomies divide-and-conquer search algorithm and
# executes in logarithmic time.


def binary_search(nlist, target):
    first = 0
    last = len(nlist)-1
    found = False

    while ( first<= last and not found ):
        mid = (first+last)//2
        if nlist[mid] == target:
            found = True
        else:
            if target < nlist[mid]:
                last = mid-1
            else:
                first = mid+1

    return found


nlist = [12,4,12,54,23,15,24,84,37,3,5]
nlist.sort()
print("list:", nlist)
target = 24
print(binary_search(nlist, target))
        

File: ruby_and_python/13_search/binary_search.rb
class Search

  def initialize(numbers, number)
    @numbers = numbers
    @number = number
  end

  def binary_search
    first = 0
    last = @numbers.size - 1

    while first <= last
      mid = ( first + last ) / 2

      return "Si" if @numbers[mid] == @number 

      if @numbers[mid] < @number
        first = mid + 1
      else
        last = mid - 1
      end
    end

    return "No"

  end
end


search = Search.new([1,4,6,9,106,305,468,563,693], 468)
#search = Search.new([1,4,6,9,106,305,468,563,693], 346)
puts search.binary_search

